src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,695 @@
+        /// the specified initial capacity for its underlying array.
+        /// </summary>
+        public PriorityQueue(int initialCapacity, IComparer<TPriority>? comparer)
+        {
+            if (initialCapacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(initialCapacity), initialCapacity, SR.ArgumentOutOfRange_NeedNonNegNum);
+            }
+
+            _nodes = new (TElement, TPriority)[initialCapacity];
+            _unorderedItems = new Lazy<UnorderedItemsCollection>(() => new UnorderedItemsCollection(this));
+            this.Comparer = comparer ?? Comparer<TPriority>.Default;
+        }
+
+        /// <summary>
+        /// Creates a priority queue populated with the specified elements and priorities.
+        /// </summary>
+        public PriorityQueue(IEnumerable<(TElement element, TPriority priority)> items)
+            : this(items, comparer: null)
+        {
+        }
+
+        /// <summary>
+        /// Creates a priority queue populated with the specified elements and priorities,
+        /// and with the specified priority comparer.
+        /// </summary>
+        public PriorityQueue(IEnumerable<(TElement element, TPriority priority)> items, IComparer<TPriority>? comparer)
+        {
+            if (items is null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            _nodes = EnumerableHelpers.ToArray(items, out _size);
+            _unorderedItems = new Lazy<UnorderedItemsCollection>(() => new UnorderedItemsCollection(this));
+            this.Comparer = comparer ?? Comparer<TPriority>.Default;
+
+            if (_size > 1)
+            {
+                Heapify();
+            }
+        }
+
+        /// <summary>
+        /// Gets the current amount of items in the priority queue.
+        /// </summary>
+        public int Count => _size;
+
+        /// <summary>
+        /// Gets the priority comparer of the priority queue.
+        /// </summary>
+        public IComparer<TPriority> Comparer { get; }
+
+        /// <summary>
+        /// Gets a collection that enumerates the elements of the queue.
+        /// </summary>
+        public UnorderedItemsCollection UnorderedItems => _unorderedItems.Value;
+
+        /// <summary>
+        /// Enqueues the specified element and associates it with the specified priority.
+        /// </summary>
+        public void Enqueue(TElement element, TPriority priority)
+        {
+            if (element is null)
+            {
+                throw new ArgumentNullException(nameof(element));
+            }
+
+            if (priority is null)
+            {
+                throw new ArgumentNullException(nameof(priority));
+            }
+
+            EnsureEnoughCapacityBeforeAddingNode();
+
+            // Add the node at the end
+            var node = (element, priority);
```

Comment:
Assigning `(element, priority)` to `node` and then passing `node` by value to `MoveUp` involves a lot of copying, which can be particularly noticeable if either `TElement` or `TPriority` are structs. I would recommend making the `MoveUp` and `MoveDown` methods [accept the values by reference](https://github.com/eiriktsarpalis/pq-tests/blob/89d53de3975fd2ef7bc217c355aadbf5f3356f10/PriorityQueue/PriorityQueue.cs#L371) to minimize the amount of copying required.

Reply:
Hmm, are you sure this is the case? I'm deep diving into the behavior now and reading https://devblogs.microsoft.com/premier-developer/the-in-modifier-and-the-readonly-structs-in-c/#performance-characteristics-of-the-in-modifier, based on which it seems that if `TElement` or `TPriority` is a normal struct (not `readonly`), then I think this would apply:

> It means that you should never pass a non-readonly struct as in parameter. It almost always will make the performance worse. Yes, the argument passing is cheaper, but once the parameter is used, the defensive copy will nullify the benefits or will make the performance worse. It could make sense if the struct is a C-like struct with a bunch of public fields and everyone in the team is aware that changing fields to properties would have a drastic performance impact on the application. But in this case, I would suggest passing the struct by reference instead.

Reply:
I wasn't aware this. TBH I hadn't benchmarked it with any mutable structs so I suppose it's probably right. But the answer really is in the paragraph you quoted, we could pass by ref instead.